Plutarch/Num.hs

```diff
@@ -0,0 +1,63 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Plutarch.Num (PNum (..), PFractional (..)) where
+
+import Plutarch.Internal (
+  PType,
+  Term,
+  pthrow,
+  punsafeCoerce,
+  (:-->),
+ )
+import Plutarch.Internal.Other (pto)
+import Plutarch.Internal.PLam ((#))
+import Plutarch.Internal.PlutusType (PInner)
+import Plutarch.Unsafe (punsafeDowncast)
+
+class PNum (a :: PType) where
```

Comment:
you know, I'm still unsure if adding a base-like `Num` is a good idea. It'd be nice to just have `Num` instances as we do already and focus on a nice, convenient numerical hierarchy instead.

this class might add backwards compat burden later on

Reply:
The reason to have this isn't to have a numerical hierarchy. It's simply to allow deriving `Num`.

Reply:
should `Semigroup`/`Monoid` get the same treatment?

Reply:
Unfortunately yeah, I'm not sure there's a way to do this for all such typeclasses. In general, `C (Term s A)` is an orphan instance, so it's not great style in the first place. For all such `C`, it seems like we want `PC a => C (Term s a)`